sample/redo/lib/main.dart

```diff
@@ -140,125 +142,122 @@ class _MyHomePageState extends State<MyHomePage> {
+      int listId = args['listId'];
+      int oldIndex = args['oldIndex'];
+      int newIndex = args['newIndex'];
+      List<Todo> todos = (await todosInListFromTx(tx, listId)).toList();
+      int id = todos[oldIndex].id;
+      double order = getNewOrder(newIndex, todos);
+      final todo = await _read(tx, id);
       if (todo == null) {
         return;
       }
       todo.order = order;
-      _write(tx, id, todo);
+      await _write(tx, todo);
     });
   }
 
-  Future<void> _handleRemove(int id) async {
-    // TODO(arv): Should be mutate
-    _replicache.query((tx) async {
-      await _del(tx, id);
-    });
-  }
+  Future<void> _addTodoItem(String text) async {
+    // Only add the task if the user actually entered something
+    if (text.isEmpty) {
+      return;
+    }
 
-  void _selectListId(int listId) {
-    setState(() {
-      _selectedListId = listId;
-    });
+    Random r = Random.secure();
+    int id = r.nextInt(1 << 32);
+    await _todoCreateMutator(
+        {'id': id, 'text': text, 'listId': _selectedListId});
   }
 
-  Future<void> _dropDatabase() async {
-    throw UnimplementedError();
-    // Navigator.pop(context);
-    // var items = await _replicache.scan(prefix: prefix);
-    // for (final ScanItem item in items) {
-    //   await _replicache.del(item.key);
-    // }
-    // await _init();
-  }
+  Future<void> _handleDone(int id, bool complete) =>
+      _todoDoneMutator({'id': id, 'complete': complete});
 
-  Future<void> _addTodoItem(String task) async {
-    var uuid = new Uuid();
-    // Only add the task if the user actually entered something
-    if (task.isNotEmpty) {
-      List<Todo> todos = _activeTodos();
-      int index = todos.isEmpty ? 0 : todos.length;
-      Random r = Random.secure();
-      int id = r.nextInt(1 << 32);
-      double order = _getNewOrder(index);
-      // TODO(arv): Should be mutate and maybe include _activeTodos.
-      _replicache.query((tx) async {
-        await _write(tx, id, Todo(id, _selectedListId, task, false, order));
-      });
-    }
-  }
+  Future<void> _handleRemove(int id) => _todoRemoveMutator({'id': id});
 
-  // calculates the order field by halving the distance between the left and right neighbor orders.
-  // min default value = -minPositive
-  // max default value = double.maxFinite
-  double _getNewOrder(int index) {
-    List<Todo> todos = _activeTodos();
-    double minOrderValue = 0;
-    double maxOrderValue = double.maxFinite;
-    double leftNeighborOrder =
-        index == 0 ? minOrderValue : todos[index - 1].order.toDouble();
-    double rightNeighborOrder =
-        index == todos.length ? maxOrderValue : todos[index].order.toDouble();
-    double order =
-        leftNeighborOrder + ((rightNeighborOrder - leftNeighborOrder) / 2);
-    return order;
-  }
+  Future<void> _handleReorder(int oldIndex, int newIndex) =>
+      _todoReorderMutator({
+        'listId': _selectedListId,
+        'oldIndex': oldIndex,
```

Comment:
Right so map `oldIndex` to an id here, then pass that into the mutator. This is important with these -- and not sure what to call the property. You want to pass in arguments that completely describe what needs to happen using nothing else other than the state of the db.

Reply:
also `newIndex` is not very good. I think what we really want is before A and after B but if "someone" else adds an item in between A and B before us the ordering is not clear.

Reply:
You are right! That was the whole point of this. What really needs to happen is that `getNewOrder` needs to come out of the transaction. Sorry that I did not see that. (The whole point is that calculated fractional indexes do basically the right thing across reorders).

In that case, my comment about ID is less important, but still might be a nice cleanup.